Transfer nested constants

myronmarston opened the pull request:
This is a followup to #4.
- This switches `fire_class_double` to use a module rather than a subclass of `RSpec::Mocks::Mock`. The reason for the change is to allow class doubles to have nested constants like real classes and modules.  I was surprised at how seamless this change was.  I think we do lose a couple things in this change, though:
  - There's not the `:__declared_as` entry in the mock options hash that rspec uses as part of its error message.  This isn't a big deal to me, though.
  - RSpec mocks support additional options when declared that aren't supported here.  Besides some deprecated things (like `:null_object => true`), I think rspec treats an options hash as a list of methods-to-stub and corresponding values.  It wouldn't be hard to add that here.
- I decided to not transfer constants by default.  I like the flexibility of being able to either have it transfer all nested constants (`:transfer_nested_constants => true`) or specify a list of constants to transfer (`:transfer_nested_constants => [:A, :B])`.  The APIs were more natural and readable this way then trying to do the inverse, I think.
- There are a few edge cases I had to think carefully about and decide what to do for.  I decided to raise errors for several of these, but I can be convinced otherwise.  Specifically, here are the cases that cause it to raise an error:
  - When `stub_const` is called with a `:transfer_nested_constants` option and the named constant is not a class or module.  The original object doesn't have any constants to transfer then.
  - When `stub_const` is called with a `:transfer_nested_constants` option and the given stub object is not a class or module.  The constants cannot be transferred to it then.
  - When a list of constants is used (`:transfer_nested_constants => [:A]`) but one or more of those are not defined constants.  I can see an argument either way on this one (especially given rspec-fire's use case of being used with as little loaded as possible), but I felt like putting a constant in the whitelist implies it is an important constant that you always want to be transferred, and thus an error to notify you that the constant is undefined is helpful.  In most cases, I would expect this to happen due to fat-fingering a constant name.
- Currently, constant transferring isn't supported when using `fire_replaced_class_double`.  Instead you have to use `fire_class_double("ClassName").as_replaced_constant(:transfer_nested_constants =>  true)`.  I felt like it would be confusing to support the option directly on `fire_replaced_class_double` since it accepts a hash of options that are intended to be test double options.  The `:transfer_nested_constants` option is specific to the constant stubbing and felt like it belonged there to keep things simpler.  This suggests to me that maybe `fire_class_double("...").as_replaced_constant` is the right interface for this.
- I noticed that the rspec-fire was broken against the just-released rspec 2.9.0.rc2, so I fixed that as well.

Feedback welcome, of course.

/cc @freelancing-god

myronmarston committed:
Deal with RSpec 2.9 renaming #mocks to #receivers.

myronmarston committed:
Use a module for a class double.

This will allow nested constants to be preserved on the class double.

myronmarston committed:
Deal with RSpec 2.9 renaming #sym to #message.

https://github.com/rspec/rspec-mocks/commit/dce2d2cd1be7609edfd1e3ae3c7b7d6cd7ac9a57

myronmarston committed:
2.9.0.rc2 is out, so consider it the latest rspec.

myronmarston committed:
Add support for transferring nested constants to a stubbed constant.

myronmarston committed:
Add support for transfer constants option to #as_replaced_constant.

myronmarston committed:
Ensure class doubles have a readable string representation.

"MyClass (fire double)" is much more readable than "#<Module:0x00000100956180>".

myronmarston committed:
Assign name to class doubles.

#name is part of the standard class interface. It's a bit silly to make people stub it.

myronmarston committed:
Raise a mock expectation error for undefined methods called on class doubles.

We got this for free when FireClassDouble subclassed RSpec::Mocks::Mock; now that we're using modules, we need to define method_missing.

xaviershay commented:
Sorry missed this PR :/

>  I think rspec treats an options hash as a list of methods-to-stub and corresponding values. It wouldn't be hard to add that here.

I want to get this added before merging so as not to break backwards compatibility (I use this feature).

> There are a few edge cases I had to think carefully about and decide what to do for.

I agree with your decisions.

> This suggests to me that maybe fire_class_double("...").as_replaced_constant is the right interface for this.

Interesting. I don't think anything needs to change on this PR as a result, happy to let it bake for a while.

myronmarston commented:
> I want to get this added before merging so as not to break backwards compatibility (I use this feature).

Sounds good, I can take a crack at that this weekend sometime.

myronmarston committed:
Support stubs being specified as a hash.

myronmarston committed:
Fix failing spec on 1.8.7.

On 1.8.7, Module#constants returns an array of strings. On 1.9 it's an array of symbols.  We need to convert the given array to strings on 1.8.7 so the two arrays can be subtracted from each other.

myronmarston commented:
There you go :).

xaviershay commented:
<img src="http://www.clipartpal.com/_thumbs/pd/education/double_gold.png">

Thanks!

myronmarston commented:
Hurray! I got a gold star!

BTW, when do you plan to cut another release?  I may blog about some of this stuff but it'd be nice if there were a gem release with it first.  I guess that solidifies the API, though...are you happy with the API?

xaviershay commented:
I'll do a release this evening. Happy with the API - worst case we have major versions for a reason :)

xaviershay commented:
Ah need to fix this first: http://travis-ci.org/#!/xaviershay/rspec-fire/builds/889802
At first glance looks related to the monkey patch I commented on?

myronmarston commented:
Yeah, I'll take a look at fixing that later tonight.
